### PR TITLE
data instead of datasets

### LINKED ISSUE #247: Change modules to use data argument instead of datasets
Followup to: 
- https://github.com/insightsengineering/teal/issues/654 
- https://github.com/insightsengineering/teal.transform/issues/61
- https://github.com/insightsengineering/teal.transform/issues/67

Required fixes:
- Remove `datasets` argument and introduce `data` argument. 
- Use new data_extract and data_merge

NOTE: teal.modules.hermes using datasets$get_data(filtered = TRUE) and datasets$get_data(filtered = FALSE). We need to resolve this situation.

### PR DESCRIPTION
**Pull request by mhallal1**
closes #247 

- `datasets` replaced by `data` in scripts and tests for Filtered and unfiltered data
- snaps are updated
- design scripts are not updated 
- PENDING: tests with `TESTING_DEPTH = 5` are not passing

### CODE REVIEW

File: R/experimentSpec.R [hunk 1]
```diff
@@ -206,13 +207,14 @@ h_gene_data <- function(object, name_annotation) {
...
-                                 datasets,
+                                 data,
+                                 filter_panel_api,
                                  mae_name,
                                  name_annotation = "symbol",
                                  sample_vars_as_factors = TRUE,
                                  with_mae_col_data = TRUE) {
   assert_string(id)
-  assert_r6(datasets)
+  assert_list(data)
```
Comment (danielinteractive) [inline R/experimentSpec.R:217]: Is there a way to make this more specific maybe? Otherwise ok as proposed

**Review by danielinteractive — COMMENTED** [review]
Thank you, I wonder if it would be better before merging this to include also passing the deep tests to be sure that everything works

File: R/experimentSpec.R (hunk 1 above)
Comment (mhallal1) [reply R/experimentSpec.R:217]: yeah this PR is not done yet, in progress

Comment (mhallal1) [reply R/experimentSpec.R:217]: I have created an issue for the deep tests failures: https://github.com/insightsengineering/teal.modules.hermes/issues/279

File: R/pca.R [hunk 2]
```diff
@@ -59,12 +59,12 @@ tm_g_pca <- function(label,
...
 ui_g_pca <- function(id,
-                     datasets,
+                     data,
                      mae_name,
                      pre_output,
                      post_output) {
   ns <- NS(id)
-  mae <- datasets$get_data(mae_name, filtered = FALSE)
+  mae <- data[[mae_name]]()
   experiment_name_choices <- names(mae)
```
Comment (mhallal1) [inline R/pca.R:68]: This code was obsolete and not used therefore it was removed.

File: R/experimentSpec.R [hunk 3]
```diff
@@ -11,15 +11,14 @@
...
                                 mae_name,
                                 label_experiments = "Select Experiment") {
   assert_string(inputId)
-  assert_r6(datasets)
+  assert_list(data)
   assert_string(mae_name, min.chars = 1L)
   assert_string(label_experiments, min.chars = 1L)
-
-  mae <- datasets$get_data(mae_name, filtered = FALSE)
+  mae <- data[[mae_name]]()
```
Comment (mhallal1) [inline R/experimentSpec.R:21]: The switch from non-filtered data to filtered data here should not have an impact as we are fetching the names of experiments only.

File: R/forestplot.R [hunk 4]
```diff
@@ -144,10 +145,13 @@ srv_g_forest_tte <- function(id,
                              plot_height,
                              plot_width) {
   with_reporter <- !missing(reporter) && inherits(reporter, "Reporter")
+  with_filter <- !missing(filter_panel_api) && inherits(filter_panel_api, "FilterPanelAPI")
```
Comment (gogonzo) [inline R/forestplot.R:148]: practically `with_filter` isn't needed as as module depends on `filter_panel_api` as `experimentSpecServer` needs it.

File: R/experimentSpec.R [hunk 5]
```diff
@@ -254,28 +255,26 @@ experimentSpecServer <- function(id,
       name <- input$name
       req(name)
 
-      filtered_mae <- datasets$get_filtered_dataset(mae_name)
-      filter_states <- filtered_mae$get_filter_states(name)
-      subset_queue <- filter_states$queue_get("subset")
-      sapply(subset_queue, function(x) x$get_call())
```
Comment (gogonzo) [inline R/experimentSpec.R:260]: this is a good change

**Review by gogonzo — COMMENTED** [review]
Changes straightforward not affecting functionality of the package. We will try to resolve deep tests problems, but these code changes doesn't seem to be a cause.

File: R/experimentSpec.R (hunk 5 above)
Comment (danielinteractive) [reply R/experimentSpec.R:260]: does this not affect functionality? will be good to run the shiny test and check that... I remember we specifically checked for the filter code here

Comment (mhallal1) [reply R/experimentSpec.R:260]: @danielinteractive is there any reason you checked the code specifically?

Comment (gogonzo) [reply R/experimentSpec.R:260]: In my opinion checking filter-code is a responsibility of teal tests not tmh

Comment (gogonzo) [reply R/experimentSpec.R:260]: @danielinteractive we test the filter-calls in several configurations in `teal.slice`, so there is no need to test them again in tmh

https://github.com/insightsengineering/teal.slice/blob/80aa3c2c6943fb41f5204edd0ff61a07b123c3c0/tests/testthat/test-MAEFilteredDataset.R#L121

https://github.com/insightsengineering/teal.slice/blob/80aa3c2c6943fb41f5204edd0ff61a07b123c3c0/tests/testthat/test-MAEFilteredDataset.R#L230

https://github.com/insightsengineering/teal.slice/blob/80aa3c2c6943fb41f5204edd0ff61a07b123c3c0/tests/testthat/test-MAEFilteredDataset.R#L301

Comment (danielinteractive) [reply R/experimentSpec.R:260]: I mean the idea was to trigger a change based on the subset code change in teal. maybe that is now handled by `teal.slice` already. I think as long as functionality does not change it is ok

Comment (gogonzo) [reply R/experimentSpec.R:260]: subset code is a consequence of the change in the "subset" filter-states. It's enough to observe if the state changed

Comment (danielinteractive) [reply R/experimentSpec.R:260]: perfect, thanks @gogonzo for confirming!

Comment (gogonzo) [reply R/experimentSpec.R:260]: Thanks @danielinteractive 
Just FYI - we will merge this PR to the feature branch and we will continue working there until we fix everything - this PR has `main` as base branch but only to have a integration tests running, eventually we will switch to feature branch and this won't be included in the nearest release.

Comment (danielinteractive) [reply R/experimentSpec.R:260]: ah ok cool thanks for info

**Review by nikolas-burkoff — COMMENTED** [review]
We can remove `MockDatasets` from tests/helper-data_funs.R

**Comment by mhallal1** [comment]
> MockDatasets

removed.

File: R/forestplot.R (hunk 4 above)
Comment (nikolas-burkoff) [reply R/forestplot.R:148]: so if module depends on filter_panel_api then we should have `stop_if_missing(filter_panel_api, "this module requires a filter panel see ....")`

Comment (mhallal1) [reply R/forestplot.R]: `with_filter` above was replaced with `checkmate::assert_class(filter_panel_api, "FilterPanelAPI` in all modules using `experimentSpecServer`.

**Review by gogonzo — APPROVED** [review]

**Pull request merged**